src/main/java/com/redhat/cloud/custompolicies/app/rest/PolicyCrudService.java

```diff
@@ -403,19 +403,42 @@ public Response updatePolicy(@QueryParam ("dry") boolean dryRun, @PathParam("pol
           return Response.status(200).entity(new Msg("Policy validated")).build();
         }
 
+        // All is good, we can now do the real work
+        // The engine requires that we update existing structures,
+        // so we need to first poll from it.
         try {
+          FullTrigger existingTrigger;
+          if (!skipEngineCall) {
+            try {
+              existingTrigger = engine.fetchTrigger(storedPolicy.id, user.getAccount());
+            } catch (Exception e) {
+              return Response.status(400, e.getMessage()).entity(getEngineExceptionMsg(e)).build();
+            }
+          }
+          else {
+            // To make the compiler happy
+            existingTrigger = new FullTrigger(storedPolicy);
+          }
+
           storedPolicy.populateFrom(policy);
           storedPolicy.customerid = user.getAccount();
-          Policy.persist(storedPolicy);
-          FullTrigger trigger = new FullTrigger(storedPolicy);
 
+          boolean okToStore;
+          existingTrigger.updateFromPolicy(storedPolicy);
           if (!skipEngineCall) {
             try {
-              engine.updateTrigger(storedPolicy.id, trigger, false, user.getAccount());
+              engine.updateTrigger(storedPolicy.id, existingTrigger, false, user.getAccount());
+              okToStore = true;
```

Comment:
Is this `okToStore` really needed? Maybe missed a case.
- If the engine failed, it would throw an exception and would return error response (thus not storing)
- If the engine succeed it would just persist
- If !skipEngine, it would just persist

Reply:
I think you are right.